lib/Dialect/SPIRV/SPIRVOps.cpp

```diff
@@ -199,6 +213,136 @@ static void printNoIOOp(Operation *op, OpAsmPrinter *printer) {
   printer->printOptionalAttrDict(op->getAttrs());
 }
 
+//===----------------------------------------------------------------------===//
+// spv.AccessChainOp
+//===----------------------------------------------------------------------===//
+
+static Type extractPtrTypeFromPtrToCompositeType(ArrayRef<Value *> indices,
+                                                 Type type, Location baseLoc) {
+  if (!indices.size()) {
+    return mlir::emitError(baseLoc, "'spv.AccessChain' op expected at least "
+                                    "one index for spv.AccessChain"),
+           nullptr;
+  }
+
+  auto ptrType = type.dyn_cast<spirv::PointerType>();
+  if (!ptrType) {
+    return mlir::emitError(baseLoc, "'spv.AccessChain' op expected a pointer "
+                                    "to composite type, but provided ")
+               << type,
+           nullptr;
+  }
+
+  auto resultType = ptrType.getPointeeType();
+  auto resultStorageClass = ptrType.getStorageClass();
+  int32_t index = 0;
+
+  for (auto indexSSA : indices) {
+    auto cType = resultType.dyn_cast<spirv::CompositeType>();
+    if (!cType) {
+      return mlir::emitError(
+                 baseLoc,
+                 "'spv.AccessChain' op cannot extract from non-composite type ")
+                 << resultType << " with index " << index,
+             nullptr;
+    }
+    index = 0;
+    if (resultType.isa<spirv::StructType>()) {
+      Operation *op = indexSSA->getDefiningOp();
+      if (!op) {
+        return mlir::emitError(baseLoc, "'spv.AccessChain' op index must be an "
+                                        "integer spv.constant to access "
+                                        "element of spv.struct"),
+               nullptr;
+      }
+
+      if (!extractValueFromConstOp(op, index)) {
+        return mlir::emitError(
+                   baseLoc,
+                   "'spv.AccessChain' index must be an integer spv.constant to "
+                   "access element of spv.struct, but provided ")
+                   << op->getName(),
+               nullptr;
+      }
+
+      if (index < 0 || static_cast<uint64_t>(index) >= cType.getNumElements()) {
+        return mlir::emitError(baseLoc, "'spv.AccessChain' op index ")
+                   << index << " out of bounds for " << type,
+               nullptr;
+      }
+    }
+    resultType = cType.getElementType(index);
+  }
+  return spirv::PointerType::get(resultType, resultStorageClass);
+}
+
+static ParseResult parseAccessChainOp(OpAsmParser *parser,
+                                      OperationState *state) {
+  OpAsmParser::OperandType ptrInfo;
+  SmallVector<OpAsmParser::OperandType, 4> indicesInfo;
+  Type type;
+  // TODO(denis0x0D): regarding to the spec an index must be any integer type,
```

Comment:
We may want to `parseOperand()` then `parseColonType()` for the operands. And iterate for all operands. This will give us a way to determine the type of each operand. But fine to leave a TODO for now.

Reply:
Thanks for mentioning this, if it's ok, I'll leave a TODO.

Reply:
SGTM!